rcl/src/rcl/arguments.c

```diff
@@ -0,0 +1,377 @@
+    void * shrunk_rules = allocator.reallocate(
+      args_impl->remap_rules, sizeof(rcl_remap_t) * args_impl->num_remap_rules, allocator.state);
+    if (NULL == shrunk_rules) {
+      ret = RCL_RET_BAD_ALLOC;
+      goto fail;
+    }
+    args_impl->remap_rules = shrunk_rules;
+  } else {
+    // No remap rules
+    allocator.deallocate(args_impl->remap_rules, allocator.state);
+    args_impl->remap_rules = NULL;
+  }
+  // Shrink unparsed_args
+  if (0 == args_impl->num_unparsed_args) {
+    // No unparsed args
+    allocator.deallocate(args_impl->unparsed_args, allocator.state);
+    args_impl->unparsed_args = NULL;
+  } else if (args_impl->num_unparsed_args < argc) {
+    void * shrunk = allocator.reallocate(
+      args_impl->unparsed_args, sizeof(int) * args_impl->num_unparsed_args, allocator.state);
+    if (NULL == shrunk) {
+      ret = RCL_RET_BAD_ALLOC;
+      goto fail;
+    }
+    args_impl->unparsed_args = shrunk;
+  }
+
+  return RCL_RET_OK;
+fail:
+  fail_ret = ret;
+  if (NULL != args_impl) {
+    // assign to ret to suppress warning about not checking return of fini
+    ret = rcl_arguments_fini(args_output, allocator);
+  }
+  return fail_ret;
+}
+
+int
+rcl_get_num_unparsed_arguments(
+  rcl_arguments_t * args)
+{
+  if (NULL == args || NULL == args->impl) {
+    return -1;
+  }
+  return args->impl->num_unparsed_args;
+}
+
+rcl_ret_t
+rcl_get_unparsed_arguments(
+  rcl_arguments_t * args,
+  rcl_allocator_t allocator,
+  int ** output_unparsed_indices)
+{
+  RCL_CHECK_ALLOCATOR_WITH_MSG(&allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(args, RCL_RET_INVALID_ARGUMENT, allocator);
+  RCL_CHECK_ARGUMENT_FOR_NULL(args->impl, RCL_RET_INVALID_ARGUMENT, allocator);
+  RCL_CHECK_ARGUMENT_FOR_NULL(output_unparsed_indices, RCL_RET_INVALID_ARGUMENT, allocator);
+
+  *output_unparsed_indices = NULL;
+  if (args->impl->num_unparsed_args) {
+    *output_unparsed_indices = allocator.allocate(
+      sizeof(int) * args->impl->num_unparsed_args, allocator.state);
+    if (NULL == *output_unparsed_indices) {
+      return RCL_RET_BAD_ALLOC;
+    }
+    for (int i = 0; i < args->impl->num_unparsed_args; ++i) {
+      (*output_unparsed_indices)[i] = args->impl->unparsed_args[i];
+    }
+  }
+  return RCL_RET_OK;
+}
+
+rcl_arguments_t
+rcl_get_zero_initialized_arguments(void)
+{
+  static rcl_arguments_t default_arguments = {
+    .impl = NULL
+  };
+  return default_arguments;
+}
+
+rcl_ret_t
+rcl_arguments_fini(
+  rcl_arguments_t * args,
+  rcl_allocator_t allocator)
+{
+  RCL_CHECK_ALLOCATOR_WITH_MSG(&allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(args, RCL_RET_INVALID_ARGUMENT, allocator);
+  RCUTILS_LOG_DEBUG_NAMED(ROS_PACKAGE_NAME, "Finalizing arguments");
+  if (args->impl) {
+    if (args->impl->remap_rules) {
+      for (int i = 0; i < args->impl->num_remap_rules; ++i) {
+        rcl_remap_fini(&(args->impl->remap_rules[i]), allocator);
```

Comment:
In 6ac891fd7a79a70e9f87926aefccbc48b379c7a3 it will log an error if it fails to finalize a remap rule, but continue trying to free the rest of the rules and the stuff in the struct. Also added `RCL_WARN_UNUSED` to `rcl_remap_fini`